File: src/main/java/pw/switchcraft/plethora/gameplay/registry/PlethoraMetaRegistration.java
package pw.switchcraft.plethora.gameplay.registry;

import net.minecraft.item.ItemStack;
import pw.switchcraft.plethora.Plethora;
import pw.switchcraft.plethora.api.meta.IMetaProvider;
import pw.switchcraft.plethora.api.meta.IMetaRegistry;
import pw.switchcraft.plethora.gameplay.BindableModuleItemMeta;
import pw.switchcraft.plethora.integration.MetaWrapper;

public class PlethoraMetaRegistration {
    public static void registerMetaProviders(IMetaRegistry r) {
        provider(r, "metaProvider", MetaWrapper.class, new MetaWrapper.MetaProvider());

        provider(r, "bindableModuleItem", ItemStack.class, new BindableModuleItemMeta());
    }

    private static <T> void provider(IMetaRegistry r, String name, Class<T> target, IMetaProvider<T> provider) {
        r.registerMetaProvider(Plethora.modId + ":" + name, Plethora.modId, target, provider);
    }
}

File: src/main/java/io/sc3/plethora/gameplay/registry/PlethoraMetaRegistration.kt
package io.sc3.plethora.gameplay.registry

import io.sc3.plethora.Plethora.modId
import io.sc3.plethora.api.meta.IMetaProvider
import io.sc3.plethora.api.meta.IMetaRegistry
import io.sc3.plethora.gameplay.modules.BindableModuleItemMeta
import io.sc3.plethora.integration.DetailsMetaWrapper.MetaProvider

object PlethoraMetaRegistration {
  @JvmStatic
  fun registerMetaProviders(registry: IMetaRegistry) {
    with (registry) {
      provider("metaProvider", MetaProvider)
      provider("bindableModuleItem", BindableModuleItemMeta)
    }
  }

  private inline fun <reified T> IMetaRegistry.provider(name: String, provider: IMetaProvider<T>) {
    registerMetaProvider("$modId:$name", modId, T::class.java, provider)
  }
}
